### PR TITLE
Dynamically resize 2

### PR DESCRIPTION
**Pull request by guibescos**
Dynamically resize stake_account_positions. Right now we only increase the size, but in a future PR we can reduce the size as well.

### CODE REVIEW

File: staking/programs/staking/src/state/positions.rs [hunk 1]
```diff
@@ -3,59 +3,152 @@ use {
...
+        let res = usize::from(*next_index);
+        *next_index = next_index
+            .checked_add(1)
+            .ok_or_else(|| error!(ErrorCode::TooManyPositions))?;
+
+        if res == position_capacity {
+            self.acc_info.realloc(
+                PositionData::LEN + usize::from(*next_index) * POSITION_BUFFER_SIZE,
+                false,
+            )?;
```
Comment (keyvankhademi) [inline staking/programs/staking/src/state/positions.rs:149]: What happens if the current len is larger? do we reduce size?

Comment (guibescos) [reply staking/programs/staking/src/state/positions.rs:149]: Future PR

File: staking/integration-tests/src/utils/account.rs [hunk 2]
```diff
@@ -58,3 +59,11 @@ pub fn fetch_account_data_bytemuck<T: Pod + Zeroable + AccountDeserialize>(
     let size = std::mem::size_of::<T>();
     T::try_deserialize(&mut &svm.get_account(account).unwrap().data.as_slice()[..size + 8]).unwrap()
 }
+
+pub fn fetch_positions_account(
```
Comment (keyvankhademi) [inline staking/integration-tests/src/utils/account.rs:63]: After merging with main the best place for this function is either in `solana/utils.rs` or `staking/utils.rs`

File: staking/programs/integrity-pool/src/lib.rs [hunk 3]
```diff
@@ -68,6 +70,7 @@ pub mod integrity_pool {
             owner: payer.to_account_info(),
             target_account: None,
             pool_authority: Some(pool_config.to_account_info()),
+            system_program: ctx.accounts.system_program.to_account_info(),
```
Comment (keyvankhademi) [inline staking/programs/integrity-pool/src/lib.rs:73]: let's keep it consistent to define variables for accounts at the top of the function.

File: staking/programs/integrity-pool/src/lib.rs [hunk 4]
```diff
@@ -132,6 +135,7 @@ pub mod integrity_pool {
             stake_account_custody:   stake_account_custody.clone(),
             target_account:          None,
             pool_authority:          Some(pool_config.to_account_info()),
+            system_program:          ctx.accounts.system_program.to_account_info(),
```
Comment (keyvankhademi) [inline staking/programs/integrity-pool/src/lib.rs:138]: same here

File: staking/programs/integrity-pool/src/state/pool.rs [hunk 5]
```diff
@@ -551,8 +549,8 @@ mod tests {
             pool_data.events[i].epoch = (i + 1) as u64;
         }
 
-        let mut positions = staking::state::positions::PositionData::default();
-        // this position should be ignored (wrong target)
+        let mut stake_positions_account = DynamicPositionArrayAccount::default();
+        let mut positions = stake_positions_account.to_dynamic_position_array(); // this position should be ignored (wrong target)
```
Comment (keyvankhademi) [inline staking/programs/integrity-pool/src/state/pool.rs:553]: I think this comment should be in another line below.

File: staking/programs/staking/src/state/positions.rs [hunk 6]
```diff
@@ -3,59 +3,156 @@ use {
...
+/// This account stores a user's positions in a dynamic sized array.
+/// Its first 40 bytes are `PositionData` (including discriminator) and the rest is a
+/// variable-length slice of `[u8; POSITION_BUFFER_SIZE]`. Each element of the array can be
+/// deserialized into an `Option<Position>`. The old invariant is maintained: For `i < next_index`,
+/// `positions[i] == Some` For `i >= next_index`, `positions[i] == None`
+/// Other invariants are that `data_len() == 40 + n * POSITION_BUFFER_SIZE` where n is an integer
+/// and that `data_len() >= 40 + next_index * POSITION_BUFFER_SIZE`
+pub struct DynamicPositionArray<'a> {
+    pub acc_info: AccountInfo<'a>, /* We store account info to get access to the data and
+                                    * resize */
```
Comment (keyvankhademi) [inline staking/programs/staking/src/state/positions.rs:43]: for consistency let's move this comment above the definition.

File: staking/programs/staking/src/state/positions.rs [hunk 7]
```diff
@@ -3,59 +3,156 @@ use {
...
+            let data = acc_info.try_borrow_mut_data()?;
+
+            // The discriminator should be zero, since we're initializing.
+            let mut disc_bytes = [0u8; 8];
+            disc_bytes.copy_from_slice(&data[..8]);
+            let discriminator = u64::from_le_bytes(disc_bytes);
+            if discriminator != 0 {
+                return Err(AnchorErrorCode::AccountDiscriminatorAlreadySet.into());
+            }
+        }
```
Comment (keyvankhademi) [inline staking/programs/staking/src/state/positions.rs:78]: what was the reason to wrap this in `{}`?

File: staking/integration-tests/src/utils/account.rs (hunk 2 above)
Comment (guibescos) [reply staking/integration-tests/src/utils/account.rs:63]: It ended up in `solana/utils.rs`

File: staking/programs/integrity-pool/src/lib.rs (hunk 3 above)
Comment (guibescos) [reply staking/programs/integrity-pool/src/lib.rs]: done

File: staking/programs/integrity-pool/src/lib.rs (hunk 4 above)
Comment (guibescos) [reply staking/programs/integrity-pool/src/lib.rs]: done

File: staking/programs/integrity-pool/src/state/pool.rs (hunk 5 above)
Comment (guibescos) [reply staking/programs/integrity-pool/src/state/pool.rs]: fixed

File: staking/programs/staking/src/state/positions.rs (hunk 6 above)
Comment (guibescos) [reply staking/programs/staking/src/state/positions.rs]: done

File: staking/programs/staking/src/state/positions.rs (hunk 7 above)
Comment (guibescos) [reply staking/programs/staking/src/state/positions.rs:78]: It borrows `account_info` so I'd have clone account_info below

**Pull request merged**